vulnerabilities/importers/github.py

```diff
@@ -0,0 +1,318 @@
+
+
+import os
+import dataclasses
+import json
+from typing import Set
+from typing import Tuple
+from typing import List
+from typing import Mapping
+from typing import Optional
+import xml.etree.ElementTree as ET
+
+import requests
+from dephell_specifier import RangeSpecifier
+from packageurl import PackageURL
+
+from vulnerabilities.data_source import Advisory
+from vulnerabilities.data_source import DataSource
+from vulnerabilities.data_source import DataSourceConfiguration
+
+
+# set of all possible values of first '%s' = {'MAVEN','COMPOSER', 'NUGET'}
+# second '%s' is interesting, it will have the value '' for the first request,
+# since we don't have any value for endCursor at the beginning
+# for all the subsequent requests it will have value 'after: "{endCursor}""
+query = """
+        query{
+        securityVulnerabilities(first: 100, ecosystem: %s, %s) {
+            edges {
+            node {
+                advisory {
+                identifiers {
+                    type
+                    value
+                }
+                summary
+                }
+                package {
+                name
+                }
+                vulnerableVersionRange
+            }
+            }
+            pageInfo {
+            hasNextPage
+            endCursor
+            }
+        }
+        }
+        """
+
+
+class GitHubTokenError(Exception):
+    pass
+
+
+@dataclasses.dataclass
+class GitHubAPIDataSourceConfiguration(DataSourceConfiguration):
+    endpoint: str
+    ecosystems: list
+
+
+class GitHubAPIDataSource(DataSource):
+
+    CONFIG_CLASS = GitHubAPIDataSourceConfiguration
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        try:
+            self.gh_token = os.environ["GH_TOKEN"]
+        except KeyError:
+            raise GitHubTokenError("Environment variable GH_TOKEN is missing")
+
+    def __enter__(self):
+        self.advisories = self.fetch()
+
+    def updated_advisories(self) -> Set[Advisory]:
+        return self.batch_advisories(self.process_response())
+
+    def fetch(self) -> Mapping[str, List[Mapping]]:
+        headers = {"Authorization": "token " + self.gh_token}
+        api_data = {}
+        for ecosystem in self.config.ecosystems:
+
+            api_data[ecosystem] = []
+            end_cursor_exp = ""
+
+            while True:
+
+                query_json = {"query": query % (ecosystem, end_cursor_exp)}
+                resp = requests.post(self.config.endpoint, headers=headers, json=query_json).json()
+                print(resp)
+                if resp.get("message") == "Bad credentials":
+                    raise GitHubTokenError("Invalid GitHub token")
+
+                end_cursor = resp["data"]["securityVulnerabilities"]["pageInfo"]["endCursor"]
+                end_cursor_exp = "after: {}".format('"{}"'.format(end_cursor))
+                api_data[ecosystem].append(resp)
```

Comment:
In general it is often better to yield rather than to accumulate in a list, FWIW